[Profiler] Avoid Recursive ThreadStoreLock in Profiling Thread Enumerator

mdh1418 opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/110062

Profiling Enumerators look to acquire the ThreadStoreLock.
In release config, re-acquiring the ThreadStoreLock and releasing it in ProfilerThreadEnum::Init will cause problems if the callback invoking EnumThread has logic that depends on the ThreadStoreLock being held.
For example, invoking `EnumThread` within `RuntimeSuspendFinished` will violate the [expectation that the ThreadStoreLock is held until RestartEE is called](https://github.com/dotnet/runtime/blob/2d6ea8de12a9f1aad9724a911d201137e5339c4b/src/coreclr/vm/threadsuspend.cpp#L5556-L5559), demonstrated in https://github.com/dotnet/runtime/issues/110062

This PR aims to avoid recursively acquiring the ThreadStoreLock by expanding the known scnearios where the profiling thread enumerator shouldn't acquire the ThreadStoreLock.

mdh1418 committed:
[Profiler] Avoid Recursive ThreadStoreLock

Profiling Enumerators look to acquire the ThreadStoreLock.
In release config, re-acquiring the ThreadStoreLock and
releasing it in ProfilerThreadEnum::Init will cause
problems if the callback invoking EnumThread has logic
that depends on the ThreadStoreLock being held.
To avoid recursively acquiring the ThreadStoreLock,
expand the condition when the profiling thread enumerator
shouldn't acquire the ThreadStoreLock.

jkotas commented:
Do we need a test for this?

VSadov commented on src/coreclr/vm/profilingenumerators.cpp:
```diff
@@ -552,9 +552,11 @@ HRESULT ProfilerThreadEnum::Init()
     // If a profiler has requested that the runtime suspend to do stack snapshots, it
     // will be holding the ThreadStore lock already
-    ThreadStoreLockHolder tsLock(!g_profControlBlock.fProfilerRequestedRuntimeSuspend);
+    ThreadStoreLockHolder tsLock(!GCHeapUtilities::IsGCInProgress() && !g_profControlBlock.fProfilerRequestedRuntimeSuspend);
```
If the intent is to not acquire thread store lock second time, then this should check if the current thread is already holding the thread store lock. 

GC could be started by other thread and GC may be changing the thread list (for example starting or background threads). 
If current thread is not holding a thread store lock, it should acquire it before iterating through thread list.

mdh1418 replied:
Thanks for pointing that out! I was mainly looking to not acquire the thread store lock a second time. Will change to `ThreadStore::HoldingThreadStore()`

VSadov commented on src/coreclr/vm/profilingenumerators.cpp:
```diff
@@ -552,9 +552,11 @@ HRESULT ProfilerThreadEnum::Init()
     // If a profiler has requested that the runtime suspend to do stack snapshots, it
     // will be holding the ThreadStore lock already
-    ThreadStoreLockHolder tsLock(!g_profControlBlock.fProfilerRequestedRuntimeSuspend);
+    ThreadStoreLockHolder tsLock(!GCHeapUtilities::IsGCInProgress() && !g_profControlBlock.fProfilerRequestedRuntimeSuspend);
```
```suggestion
    ThreadStoreLockHolder tsLock(!ThreadStore::HoldingThreadStore() && !g_profControlBlock.fProfilerRequestedRuntimeSuspend);
```

VSadov replied:
I've also looked at uses of `fProfilerRequestedRuntimeSuspend` and I think there are couple races around setting and clearing this flag.

The flag should be set _after_ the profiler succeeded at suspending the runtime, not before. 
https://github.com/dotnet/runtime/blob/016d356e43659b7789d23defe570628c8864ee73/src/coreclr/vm/proftoeeinterfaceimpl.cpp#L6838-L6839

And the flag should be cleared _before_ resuming the runtime, not after. 
Otherwise if resuming thread is preempted right after it resumes the runtime, the flag could stay set for a very long time, while the threadstore is no longer owned by the profiler.
https://github.com/dotnet/runtime/blob/016d356e43659b7789d23defe570628c8864ee73/src/coreclr/vm/proftoeeinterfaceimpl.cpp#L6877-L6878

Perhaps fix these races in the same change?

mdh1418 replied:
I can swap the order in which `fProfilerRequestedRuntimeSuspend` is set. I don't think we have anything testing its behavior very well. Wouldn't setting the flag after `SuspendEE` returns also mean that any callback during `SuspendEE` wouldn't see that it was a profiler requested suspend? (which would be fine for this scenario as it would see that the ThreadStoreLock was already acquired)

VSadov replied:
> Wouldn't setting the flag after SuspendEE returns also mean that any callback during SuspendEE wouldn't see that it was a profiler requested suspend?

I think until the suspend succeeded, it is not useful for other threads to know that profiler is trying to do it. At least not as a way to tell if it is safe to iterate the thread list. 
For the thread that is performing the suspension checking if it holds the lock is a more direct way to tell when it is ok to iterate. 

Perhaps we have other reasons why the order does not matter, but setting the flag before holding the lock seems at best confusing.

davmason replied:
It shouldn't matter to change the order. The only impact I can think of is that a profiler might get an error for a thread that has just been suspended, but a profiler should be calling SuspendRuntime and only starting sampling once that method has returned.

mdh1418 replied:
Updated and swapped the order!

mdh1418 committed:
[Profiler] Change order to set fProfilerRequestedRuntimeSuspend

There was a potential race condition when setting the flag
before suspending and resetting the flag after restarting.
For example, if the thread restarting runtime is preempted
right after resuming runtime, the flag could remain unset
by the time another thread looks to suspend runtime, which
would see that the flag as set.

mdh1418 committed:
[Profiler][Tests] Add unit test for EnumThreads during suspension

davmason reviewed (approved):
LGTM

mdh1418 committed:
[Profiler][Tests] Fixup EnumThreads test

mdh1418 commented:
Given that this test should run quickly when it passes, and its expected failure case is when it hangs/deadlocks, is there a straightforward way to set a quick timeout for this test instead of hitting CI's timeouts? And is there a way to hit such a timeout in a local run via corerun? @davmason @jkotas

davmason commented:
The problem with timeouts is you can't check them in because we run a lot of tests in various modes (GCStress, JITStress, etc) that can make them take 10+ minutes. For this kind of thing what I have done in the past is make the timeout short locally and run it in a loop on your dev machine to build confidence, but leave the timeout unchanged in the tree.

mdh1418 commented:
/backport to release/9.0-staging
